File: src/packages/sdk/src/collector.ts
/**
 * SDK helper for reporting traces to the collector endpoint.
 *
 * Usage:
 *   import { reportTrace } from '@ai-eval-platform/sdk';
 *   await reportTrace(client, { ... });
 *
 * Supports:
 *   - Client-side sampling (sampleRate)
 *   - Error traces always sent
 *   - Thumbs-down feedback always sent
 */

import type { AIEvalClient } from "./client";

export interface CollectorSpanInput {
	span_id: string;
	type?: "llm" | "tool" | "agent" | "retrieval" | "default";
	name: string;
	parent_span_id?: string;
	input?: unknown;
	output?: unknown;
	model?: string;
	vendor?: string;
	params?: Record<string, unknown>;
	metrics?: {
		prompt_tokens?: number;
		completion_tokens?: number;
		total_time_ms?: number;
	};
	timestamps?: {
		started_at: number;
		finished_at: number;
	};
	error?: { message: string; code?: string } | null;
	behavioral?: Record<string, unknown>;
	metadata?: Record<string, unknown>;
}

export interface CollectorFeedbackInput {
	type: "thumbs_up" | "thumbs_down" | "rating" | "comment";
	value?: unknown;
	user_id?: string;
}

export interface ReportTraceInput {
	trace_id: string;
	name: string;
	status?: "pending" | "success" | "error";
	duration_ms?: number;
	source?: "sdk" | "api" | "cli";
	environment?: "production" | "staging" | "dev";
	metadata?: Record<string, unknown>;
	spans: CollectorSpanInput[];
	user_feedback?: CollectorFeedbackInput;
}

export interface ReportTraceOptions {
	/** Client-side sample rate (0-1). Default: 1.0 (send all). */
	sampleRate?: number;
}

export interface ReportTraceResult {
	sent: boolean;
	trace_id: string;
	trace_db_id?: number;
	span_count?: number;
	queued_for_analysis?: boolean;
	skip_reason?: "sampled_out";
}

/**
 * Report a production trace to the collector endpoint.
 *
 * Client-side sampling: set `options.sampleRate` (0-1).
 * Error traces and thumbs-down feedback bypass sampling.
 */
export async function reportTrace(
	client: AIEvalClient,
	input: ReportTraceInput,
	options: ReportTraceOptions = {},
): Promise<ReportTraceResult> {
	const { sampleRate = 1.0 } = options;

	// Client-side sampling — errors and thumbs-down always pass through
	const isError = input.status === "error";
	const isNegativeFeedback = input.user_feedback?.type === "thumbs_down";
	const bypassSampling = isError || isNegativeFeedback;

	if (!bypassSampling && sampleRate < 1.0) {
		if (Math.random() >= sampleRate) {
			return {
				sent: false,
				trace_id: input.trace_id,
				skip_reason: "sampled_out",
			};
		}
	}

	// Use the client's internal fetch to POST to /api/collector
	const response = await (client as any).request("/api/collector", {
		method: "POST",
		body: JSON.stringify(input),
	});

	return {
		sent: true,
		trace_id: input.trace_id,
		trace_db_id: response.trace_db_id,
		span_count: response.span_count,
		queued_for_analysis: response.queued_for_analysis,
	};
}

File: src/packages/sdk-python/src/evalgate_sdk/collector.py
"""Production trace collector — report traces with client-side sampling.

Port of the TypeScript SDK's ``collector.ts``.

Usage::

    from evalgate_sdk import AIEvalClient
    from evalgate_sdk.collector import report_trace

    client = AIEvalClient(api_key="...")
    result = await report_trace(client, ReportTraceInput(
        trace_id="t-123",
        name="chat-completion",
        spans=[CollectorSpanInput(span_id="s-1", name="llm-call")],
    ))
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import Any, Literal


@dataclass
class CollectorSpanInput:
    span_id: str
    name: str
    type: Literal["llm", "tool", "agent", "retrieval", "default"] | None = None
    parent_span_id: str | None = None
    input: Any = None
    output: Any = None
    model: str | None = None
    vendor: str | None = None
    params: dict[str, Any] | None = None
    metrics: dict[str, Any] | None = None
    timestamps: dict[str, float] | None = None
    error: dict[str, Any] | None = None
    behavioral: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"span_id": self.span_id, "name": self.name}
        for key in (
            "type",
            "parent_span_id",
            "input",
            "output",
            "model",
            "vendor",
            "params",
            "metrics",
            "timestamps",
            "error",
            "behavioral",
            "metadata",
        ):
            val = getattr(self, key)
            if val is not None:
                d[key] = val
        return d


@dataclass
class CollectorFeedbackInput:
    type: Literal["thumbs_up", "thumbs_down", "rating", "comment"]
    value: Any = None
    user_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"type": self.type}
        if self.value is not None:
            d["value"] = self.value
        if self.user_id is not None:
            d["user_id"] = self.user_id
        return d


@dataclass
class ReportTraceInput:
    trace_id: str
    name: str
    spans: list[CollectorSpanInput] = field(default_factory=list)
    status: Literal["pending", "success", "error"] | None = None
    duration_ms: float | None = None
    source: Literal["sdk", "api", "cli"] | None = None
    environment: Literal["production", "staging", "dev"] | None = None
    metadata: dict[str, Any] | None = None
    user_feedback: CollectorFeedbackInput | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "trace_id": self.trace_id,
            "name": self.name,
            "spans": [s.to_dict() for s in self.spans],
        }
        for key in ("status", "duration_ms", "source", "environment", "metadata"):
            val = getattr(self, key)
            if val is not None:
                d[key] = val
        if self.user_feedback is not None:
            d["user_feedback"] = self.user_feedback.to_dict()
        return d


@dataclass
class ReportTraceOptions:
    """Options for ``report_trace``."""

    sample_rate: float = 1.0


@dataclass
class ReportTraceResult:
    sent: bool
    trace_id: str
    trace_db_id: int | None = None
    span_count: int | None = None
    queued_for_analysis: bool | None = None
    skip_reason: str | None = None


async def report_trace(
    client: Any,
    input: ReportTraceInput,
    options: ReportTraceOptions | None = None,
) -> ReportTraceResult:
    """Report a production trace to the collector endpoint.

    Client-side sampling: set ``options.sample_rate`` (0–1).
    Error traces and thumbs-down feedback bypass sampling.
    """
    opts = options or ReportTraceOptions()

    is_error = input.status == "error"
    is_negative_feedback = input.user_feedback is not None and input.user_feedback.type == "thumbs_down"
    bypass_sampling = is_error or is_negative_feedback

    if not bypass_sampling and opts.sample_rate < 1.0 and random.random() >= opts.sample_rate:
        return ReportTraceResult(
            sent=False,
            trace_id=input.trace_id,
            skip_reason="sampled_out",
        )

    try:
        response = await client._request("POST", "/api/collector", json=input.to_dict())
    except Exception as exc:
        return ReportTraceResult(
            sent=False,
            trace_id=input.trace_id,
            skip_reason=f"request_failed: {exc}",
        )

    return ReportTraceResult(
        sent=True,
        trace_id=input.trace_id,
        trace_db_id=response.get("trace_db_id"),
        span_count=response.get("span_count"),
        queued_for_analysis=response.get("queued_for_analysis"),
    )
